Possible simplifications to servant example

mchowning opened the pull request:
Took a shot at seeing how much I could trim down Main.hs to make it show as little beyond servant as possible for demonstrative purposes.  This probably wouldn't be good finished code for us to use, but I think that something along these lines could be good to use to show what servant alone, divorced from any persistence concerns, looks like.  Then made similar changes to InMemoryDb in an effort to make comparing Main.hs and InMemoryDb.hs more straightforward.

jasonstolaruk commented on src/InitData.hs:
```diff
@@ -6,13 +6,12 @@ import Types
+vehicles :: [Vehicle]
+vehicles = [ Vehicle "vin0" 2016 "M.Plus" []
+           , Vehicle "vin1" 2015 "Void"   [ Issue Battery    Low
+                                                   , Issue Electrical High
```
Could you horizontally align these 2 lines with the opening square bracket above?

mchowning replied:
Done

jasonstolaruk commented on servant-back-end.cabal:
```diff
@@ -15,16 +15,16 @@ executable servant-back-end
-                       , transformers >=0.4 && <0.5
-                       , wai >=3.0 && <4.0
-                       , warp >=3.1 && <4.0
+  build-depends:         base 
```
It's fine to remove the version constraints for now, but we should put them back in (based on what stack has chosen) before publishing the blog post. The version constraints are nice to have for people who are not using stack.

mchowning replied:
good point

mchowning replied:
Added the version constraints

jasonstolaruk commented on src/InMemoryDb.hs:
```diff
@@ -54,39 +54,47 @@ server ior = getAllVehicles
-                    v'    = v { dbId = Just newId }
-                in (IM.insert newId v' tbl, v')
-    -- echo '{"color":"red","year":2013,"model":"Void","issues":[{"issueType":"Electrical","priority":"High"}],"vin":"vin x"}' | curl -X POST -d @- http://localhost:8081/vehicles --header "Content-Type:application/json"
+        insert :: IM.IntMap Vehicle -> (IM.IntMap Vehicle, Vehicle)
```
I was slightly confused by the name "insert" due to it already being used by the container libraries. Would you mind changing it to something a little different ("insertTbl" for example)?

mchowning replied:
good idea

mchowning replied:
Done

jasonstolaruk commented on src/InMemoryDb.hs:
```diff
@@ -54,39 +54,47 @@ server ior = getAllVehicles
-                in (IM.insert newId v' tbl, v')
-    -- echo '{"color":"red","year":2013,"model":"Void","issues":[{"issueType":"Electrical","priority":"High"}],"vin":"vin x"}' | curl -X POST -d @- http://localhost:8081/vehicles --header "Content-Type:application/json"
+        insert :: IM.IntMap Vehicle -> (IM.IntMap Vehicle, Vehicle)
+        insert tbl = let newUniqueId = head . ([0..] \\) $ IM.keys tbl
```
Dotz had reservations about ```([0..] \\)``` due to its potential to confound Haskell newbies. Would you like to suggest a more readable way to do this?

mchowning replied:
I wasn't able to think of anything better without getting a lot more verbose when I was doing this.  I was hoping that adding "unique" to the variable name would help reveal the intention to make it easier to understand (or better yet, to make it clear that you don't really need to understand what is going on there).  I'll give it some thought though and see if I can come up with something better.

jasonstolaruk commented on src/InMemoryDb.hs:
```diff
@@ -54,39 +54,47 @@ server ior = getAllVehicles
-    -- echo '{"color":"grey","year":2012,"model":"Iterate","issues":[{"issueType":"Exhaust","priority":"Low"}],"vin":"vin y"}' | curl -X PUT -d @- http://localhost:8081/vehicles/0 --header "Content-Type:application/json"
+        f :: IM.IntMap Vehicle -> (IM.IntMap Vehicle, Maybe Vehicle)
+        f tbl = if IM.member i tbl
+                then let updatedTbl = IM.insert i v tbl
```
Is there a reason you chose ```updatedTbl``` over ```tbl'```? Just curious.

mchowning replied:
I thought it was a bit more descriptive.  I do not feel remotely strongly about this change though.

mchowning replied:
I went ahead and switched all the ```updatedTbl``` to ```tbl'```.  After thinking about it some more, I think that makes sense.

mchowning commented:
@jasonstolaruk Added a couple of commits addressing your comments.  Also tweaked the commit messages on a couple of the older commits, so that's why it looks like I did more than add two commits.

jasonstolaruk commented on servant-back-end.cabal:
```diff
@@ -16,9 +16,9 @@ executable servant-back-end
   other-extensions:    DataKinds, DeriveGeneric, OverloadedStrings, TypeOperators
   build-depends:         base >=4.8 && <4.9
-                       , containers >=0.5 && <0.6
+                       , containers >=0.5 && < 0.6
```
This is rather nitpicky, but since this code is for the blog post, I'm bringing it up. I don't think we need the changes on lines 19 and 21, as they break the conventional style for the "build-depends" field.

mchowning replied:
No it's a good point.  Just an oversight on my part.

jasonstolaruk commented on src/InMemoryDb.hs:
```diff
@@ -54,39 +54,47 @@ server ior = getAllVehicles
-                in (IM.insert newId v' tbl, v')
-    -- echo '{"color":"red","year":2013,"model":"Void","issues":[{"issueType":"Electrical","priority":"High"}],"vin":"vin x"}' | curl -X POST -d @- http://localhost:8081/vehicles --header "Content-Type:application/json"
+        insertIntoTbl :: IM.IntMap Vehicle -> (IM.IntMap Vehicle, Vehicle)
+        insertIntoTbl tbl = let newUniqueId = head . ([0..] \\) $ IM.keys tbl
```
Were you going to come up with something more straightforward than ```head . ([0..] \\)```?
